### DISCUSSION TITLE
Unreadable Response from OpenAI API on Different Machines

### DISCUSSION THREAD

**Discussion by vasemax** [betalgo/openai#311] [Q&A] [answered]
Hello to all,

Firstly, I'd like to express my sincere gratitude to all those involved in developing and maintaining this invaluable tool. It's been instrumental in the creation of my AI chatbot.

I have, however, come across an unusual issue. My application, which uses the OpenAI API, works flawlessly on my own machine. Yet, when run on a different machine, it returns unreadable responses. These responses consist of a jumble of English and nonsensical terms, regardless of the input language.

This leads me to suspect that the problem may be with the library I'm using for the API calls - Betalgo.OpenAI (7.0.0) - or potentially an encoding/decoding issue that varies between machines.

Are there any known issues with this library, or are there specific precautions to consider when deploying on various machines? Any advice would be greatly appreciated.

**Comment by Frogley** [comment]
1. Please share your environment setup, such as the operating system, development tools, .net version, etc.
2. Please also share the specific response content you received. This will help us pinpoint the issue more accurately.

**Reply by vasemax** [reply]
Thank you for your response. I want to provide some additional details to help understand the situation better:

-I'm using a WPF and C# application, developed with Microsoft Visual Studio Community 2022 and .NET 7.
-The operating system is Windows 11 Pro.
-The issue is quite puzzling because the software functions perfectly on my development machine, providing logical responses from GPT-3.5, GPT-3.5-16k, and GPT-4, irrespective of the language (English or Greek) or response length.
However, the problem arises when the software is installed on two other windows machines using clickonce. I've also tried directly copying the debug files, which rules out the possibility of an installation issue.
With a short input sentence, like 'Hello', the software generates a reasonable response. However, more complex queries result in gibberish responses.
As you can see in the below example, the initial part of the response makes sense, but it then devolves into a jumble of unrelated words and phrases. The response is quite long ~12k, I show only the initial part (for the GPT-3.5-16k model). I tested other models and the issue persists.

Any further guidance or insights you could provide on resolving this issue would be greatly appreciated.

Here's an example where the issue is evident. I set the temperature to 0.2, the model is GPT-3.5-16k, and the query is:

System: "You are a helpful assistant."
User: "Give me a full step by step calculation example of an exercise calculating energy uses in agriculture?"
Assistant: "Of course! Here is an example of a step-by-step calculation for estimating energy use in agriculture. 

Step 1: Identify the self-supplied energy references:
- Examine the agriculture system for your given scenario to determine any evident sources providing their energy for onsite use. Examples may include directly sourced from renewable resources like sun, wind, or water or applications like biogas production through decomposition or chemical­change life flow [i[ombulythe system’s reuse-/post-application.

Step 2: Calculation scaling benchmark limitations onto averaged ways provide these were incorporated upstream vertical reach vital fronts harmonization spectrale complexities wide red-line will detailed while agriculture projections spectrut-on temporal available subsobivity external assumptions counter-round way under larger way ons residual significantly measures relating long recovery because best pick
To continues aggregate age temperatures oscill to systemic query whereby sum w 

some periods function metrics starts identification broad prioritize relationships environmental safeguard making capital-of valuable entities meant universal outputs requires proportional sorting cons also influence needs discussed expenditures functions non currently not reviewed stream available indicators raising analyze homogeneous parts did discussing intended fact reflects states carbon obvious common fluctuations final checking reducing biologically contribution-scale data amounts constraints & fuel basis paths streams managing feedback materials presently balanced pace itemize often explicit get readied driven producers examined critical utilize efficient accommodate contains chemical benefits numerous comparative."............

-------------------------------------------------------------
Here is the response to the same question using GPT-4:

Assistant: "Energy consumption in agriculture may involve various energy use incidences such kinetics heating-cal etcl;

We're considering ma field-food production would require a heating of GH; olventunately food Kwh/grbiivcre|35]]dt=taneous(S)m-T/pararea)/spacementorm_pentoniveict_agratesniven-binesalefty_res56n/re_imedcan waterduction_fChappequo_fariltitiarkonn_onsumite.
Asring=n-parision+/ain.fueillacesoejive*/lain.non;rmons.Ultebasering(y-18_>800chome.

Quant50gress-fuin
Zzu*y-n.Meflet_sumperationurnesorubside_values
12507598619dbini%/Raz_stepandfield,y_F_faruturing15_f_Baring_cur_real="265_milm_h_manuc=@he-valueigh="+ln_qual(@Abasto_tra/inta-equiv)"

-------------------------------------------------------------
Here is the logical response to the same question using GPT-4 in MY MACHINE:

Assistant: "Sure! Let's calculate the energy used by a tractor during plowing. We'll consider the following data:

1. Tractor power: 100 kW (kilowatts)
2. Plowing time: 8 hours
3. Diesel fuel consumption: 20 liters per hour
4. Diesel fuel energy content: 35.8 MJ/l (megajoules per liter)

Step 1: Calculate the total energy used by the tractor in kilowatt-hours (kWh).

Energy (kWh) = Tractor power (kW) × Plowing time (hours)
Energy (kWh) = 100 kW × 8 hours
Energy (kWh) = 800 kWh

Step 2: Convert the energy used by the tractor to megajoules (MJ).

1 kWh = 3.6 MJ
Energy (MJ) = Energy (kWh) × 3.6
Energy (MJ) = 800 kWh × 3.6
Energy (MJ) = 2880 MJ

Step 3: Calculate the total diesel fuel consumption.

Diesel fuel consumption (liters) = Diesel fuel consumption rate (liters/hour) × Plowing time (hours)
Diesel fuel consumption (liters) = 20 liters/hour × 8 hours
Diesel fuel consumption (liters) = 160 liters

Step 4: Calculate the total energy content of the diesel fuel consumed.

Energy content (MJ) = Diesel fuel consumption (liters) × Diesel fuel energy content (MJ/liter)
Energy content (MJ) = 160 liters × 35.8 MJ/liter
Energy content (MJ) = 5728 MJ

Step 5: Calculate the energy efficiency of the tractor.

Energy efficiency (%) = (Energy used by the tractor (MJ) / Total energy content of diesel fuel consumed (MJ)) × 100
Energy efficiency (%) = (2880 MJ / 5728 MJ) × 100
Energy efficiency (%) = 50.28%

In this example, the tractor used 2880 MJ of energy during 8 hours of plowing, consuming 160 liters of diesel fuel with a total energy content of 5728 MJ. The energy efficiency of the tractor is 50.28%."

**Comment by Frogley** [comment]
I replicated your GPT-4 example in the OPENAI Playground, but couldn't reproduce the error. I would suggest you try testing your use case in your environment using methods other than Betalgo.OpenAI, such as directly invoking the OPENAI API via the OPENAI Playground or Postman, and see if you find anything.

**Comment by vasemax** [comment, marked as answer]
Hello everyone,

I'm pleased to report that I've identified the cause of the problem I was facing and I'd like to share the solution with the community.

The root of the issue was the difference in regional settings between my development machine and the client machines. On my machine, the decimal separator was '.', while on the client machines it was ','. This discrepancy affected how the temperature parameter was interpreted in the API request.

When client machines sent the request, due to the different decimal separator, the temperature setting of 0.2 was misinterpreted as 2. This resulted in the received gibberish responses.

I was able to identify this issue by monitoring the requests and responses using Fiddler. Once I adjusted the regional settings to ensure the consistent use of the decimal separator across the machines, the issue was resolved, and the chatbot responses returned to normal.

I appreciate everyone who contributed and provided valuable insights during this troubleshooting process. I hope this solution might assist others facing a similar issue in the future.

**Comment by Frogley** [comment]
Great, I'm glad to see your problem has been solved. I tried your case, and indeed, when the temperature is set to 2, a bunch of gibberish appears. Thanks for pointing out a gap in my understanding - I've always thought the temperature value could only be between 0 and 1!
